sound/soc/sof/ipc.c

```diff
@@ -634,27 +643,95 @@ int snd_sof_ipc_set_comp_data(struct snd_sof_ipc *ipc,
 					cdata->num_elems);
 
 	} else {
-		/* write value via slower IPC */
+		/* calculate header and data size */
+		switch (ctrl_type) {
+		case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+		case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+			msg_bytes = scontrol->num_channels *
+				sizeof(struct sof_ipc_ctrl_value_chan);
+			hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+			break;
+		case SOF_CTRL_TYPE_VALUE_COMP_GET:
+		case SOF_CTRL_TYPE_VALUE_COMP_SET:
+			msg_bytes = scontrol->num_channels *
+				sizeof(struct sof_ipc_ctrl_value_comp);
+			hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+			break;
+		case SOF_CTRL_TYPE_DATA_GET:
+		case SOF_CTRL_TYPE_DATA_SET:
+			msg_bytes = cdata->data->size;
+			hdr_bytes = sizeof(struct sof_ipc_ctrl_data) +
+				sizeof(struct sof_abi_hdr);
+			break;
+		default:
+			dev_dbg(ipc->sdev->dev, "ctrl type not supported %u\n",
+				ctrl_type);
+			return -EINVAL;
+		}
+
+		total_bytes = msg_bytes + hdr_bytes;
+
 		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
 		cdata->cmd = ctrl_cmd;
 		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
+		cdata->rhdr.hdr.size = total_bytes;
+		cdata->msg_id = 0;
 		cdata->comp_id = scontrol->comp_id;
 		cdata->num_elems = scontrol->num_channels;
 
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to set control %d values\n",
-				cdata->comp_id);
-			return err;
+		/* send normal size ipc */
+		if (total_bytes <= SOF_IPC_MSG_MAX_SIZE)
+			return sof_ipc_tx_message(sdev->ipc,
+						  cdata->rhdr.hdr.cmd, cdata,
+						  cdata->rhdr.hdr.size,
+						  cdata, cdata->rhdr.hdr.size);
+
+		/* if bigger than max ipc size, chop to smaller pieces */
+		dev_dbg(ipc->sdev->dev,
+			"large ipc size %u, control size %u rhdr size %u\n",
+			total_bytes, scontrol->size, cdata->rhdr.hdr.size);
+
+		/* calculate payload size and number of messages */
+		size = SOF_IPC_MSG_MAX_SIZE - hdr_bytes;
+		num_msg = (msg_bytes + size - 1) / size;
+
+		/* allocate max ipc size because we have at least one */
+		hdr = (struct sof_ipc_ctrl_data *)
+			kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+		if (!hdr)
+			return -ENOMEM;
+
+		/* copy the header data */
+		memcpy(hdr, cdata, hdr_bytes);
+
+		/* copy the data in a loop */
+		for (i = 0; i < num_msg; i++) {
+			hdr->msg_id = i;
+			send_bytes = msg_bytes > size ? size : msg_bytes;
+			hdr->rhdr.hdr.size = hdr_bytes + send_bytes;
+			if (i > 0)
+				hdr->data->size = send_bytes;
```

Comment:
I couldn't figure out the value used for i==0?

Reply:
yes it is not very clear. for i = 0 the value is the original (i.e total) size of the blob and fw will use that to reserve the memory block. I can make it more clear if I can use the abi_hdr reserved values for example (i could put total_size, current_size and possible msg_coung and msg_id.)